### PR TITLE
 InstancePool: min-available support

### PR DESCRIPTION
**Pull request by tgrondier**
# Description

Under the hood this is a full port of instancepool management and commands from egoscale v2 to egoscale v3!

## Checklist
(For exoscale contributors)

* [x] Changelog updated (under *Unreleased* block)
* [x] Acceptance tests OK
* [x] For a new resource, datasource or new attributes: acceptance test added/updated

## Testing

Acceptance tests:

```
TF_ACC=1 /usr/local/go/bin/go test                      \
        -race                   \
        -timeout=90m            \
        -tags=testacc           \
        -v -parallel=3 -count=1 -failfast   \
        github.com/exoscale/terraform-provider-exoscale/pkg/resources/instance_pool
=== RUN   TestInstancePool
=== RUN   TestInstancePool/DataSource
=== RUN   TestInstancePool/DataSourceList
=== RUN   TestInstancePool/Resource
--- PASS: TestInstancePool (167.55s)
    --- PASS: TestInstancePool/DataSource (57.13s)
    --- PASS: TestInstancePool/DataSourceList (41.89s)
    --- PASS: TestInstancePool/Resource (68.53s)
PASS
ok      github.com/exoscale/terraform-provider-exoscale/pkg/resources/instance_pool     168.582s
```

### CODE REVIEW

File: pkg/resources/instance_pool/datasource.go [hunk 1]
```diff
@@ -186,11 +190,18 @@ func dsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.
 
 	zone := d.Get(AttrZone).(string)
 
-	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutRead))
-	ctx = exoapi.WithEndpoint(ctx, exoapi.NewReqEndpoint(config.GetEnvironment(meta), zone))
+	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutCreate))
```
Comment (kobajagi) [inline pkg/resources/instance_pool/datasource.go]:
```suggestion
	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutRead))
```

File: pkg/resources/instance_pool/datasource.go [hunk 2]
```diff
@@ -205,58 +216,68 @@ func dsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.
...
+		)
+	} else {
+		rep, err := client.ListInstancePools(ctx)
+		if err == nil {
+			return diag.FromErr(err)
+		}
+		for _, p := range rep.InstancePools {
+			if p.Name == name {
+				pool = &p
+				break
```
Comment (kobajagi) [inline pkg/resources/instance_pool/datasource.go]: There is also FindInstancePool helper: https://github.com/exoscale/egoscale/blob/master/v3/operations.go#L10016

File: pkg/resources/instance_pool/resource.go [hunk 3]
```diff
@@ -448,89 +483,122 @@ func rUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag
 
 	zone := d.Get(AttrZone).(string)
 
-	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutUpdate))
-	ctx = exoapi.WithEndpoint(ctx, exoapi.NewReqEndpoint(config.GetEnvironment(meta), zone))
+	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutCreate))
```
Comment (kobajagi) [inline pkg/resources/instance_pool/resource.go]:
```suggestion
	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutUpdate))
```

File: pkg/resources/instance_pool/datasource.go (hunk 1 above)
Comment (tgrondier) [reply pkg/resources/instance_pool/datasource.go]: Additional fixes in cd27a3e8

File: pkg/resources/instance_pool/resource.go [hunk 4]
```diff
@@ -415,18 +443,25 @@ func rRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.D
 
 	zone := d.Get(AttrZone).(string)
 
-	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutRead))
-	ctx = exoapi.WithEndpoint(ctx, exoapi.NewReqEndpoint(config.GetEnvironment(meta), zone))
+	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutCreate))
```
Comment (kobajagi) [inline pkg/resources/instance_pool/resource.go]:
```suggestion
	ctx, cancel := context.WithTimeout(ctx, d.Timeout(schema.TimeoutRead))
```

File: pkg/resources/instance_pool/datasource.go (hunk 2 above)
Comment (tgrondier) [reply pkg/resources/instance_pool/datasource.go]: Fixed in [c3c4b9d](https://github.com/exoscale/terraform-provider-exoscale/pull/406/commits/c3c4b9dc5f133e03abe934db45ef4ab939dede1c)

**Review by sauterp — APPROVED** [review]

**Review by kobajagi — APPROVED** [review]

**Pull request merged**